mllib/src/main/scala/org/apache/spark/mllib/tree/impl/DecisionTreeMetadata.scala

```diff
@@ -183,11 +183,40 @@ private[spark] object DecisionTreeMetadata extends Logging {
         }
       case _ => featureSubsetStrategy
     }
+
+    object featureSubsetNumber {
+      def unapply(strategy: String): Option[Int] = try {
+        val number = strategy.toInt
+        if (0 < number) {
+          Some(number)
+        } else {
+          None
+        }
+      } catch {
+        case _ : java.lang.NumberFormatException => None
+      }
+    }
+
+    object featureSubsetFraction {
+      def unapply(strategy: String): Option[Double] = try {
+        val fraction = strategy.toDouble
+        if (0.0 < fraction && fraction <= 1.0) {
+          Some(fraction)
+        } else {
+          None
+        }
+      } catch {
+        case _ : java.lang.NumberFormatException => None
+      }
+    }
+
     val numFeaturesPerNode: Int = _featureSubsetStrategy match {
       case "all" => numFeatures
       case "sqrt" => math.sqrt(numFeatures).ceil.toInt
       case "log2" => math.max(1, (math.log(numFeatures) / math.log(2)).ceil.toInt)
       case "onethird" => (numFeatures / 3.0).ceil.toInt
+      case featureSubsetNumber(number) => if (number > numFeatures) numFeatures else number
+      case featureSubsetFraction(fraction) => (fraction * numFeatures).ceil.toInt
```

Comment:
Again, I think regex can simplify things a bit here. I haven't tried the following, but it should work.

```scala
val isIntRegex = "(\\d+)".r
val isFractionRegex = "(0?\\.\\d+|1.0+)".r
val numFeaturesPerNode: Int = _featureSubsetStrategy match {
      case "all" => numFeatures
      case "sqrt" => math.sqrt(numFeatures).ceil.toInt
      case "log2" => math.max(1, (math.log(numFeatures) / math.log(2)).ceil.toInt)
      case "onethird" => (numFeatures / 3.0).ceil.toInt
      case isIntRegex(number) => if (number > numFeatures) numFeatures else number
      case isFractionRegex(fraction) => (fraction * numFeatures).ceil.toInt
    }
```

Then we could avoid adding the `featureSubset*` objects.